tasty-plutus/src/Test/Tasty/Plutus/Internal/Monad.hs

```diff
@@ -0,0 +1,337 @@
+  vcat,
+  ($+$),
+  (<+>),
+  (<>),
+ )
+import Text.Show.Pretty (ppDoc, valToDoc)
+import Prelude hiding ((<>))
+
+data SomeScript (p :: Purpose) where
+  SomeSpender ::
+    Validator ->
+    SomeScript 'ForSpending
+  SomeMinter ::
+    MintingPolicy ->
+    SomeScript 'ForMinting
+
+class
+  Monad m =>
+  MonadScriptTesting (m :: Type -> Type) (p :: Purpose) (r :: Type)
+    | m -> p
+    , m -> r
+  where
+  getShouldSucceed :: m Bool
+  getMPred :: m (Maybe (Vector Text -> Bool))
+  getCB :: m (ContextBuilder p)
+  getTestData :: m (TestData p)
+  getScript :: m (SomeScript p)
+  getOpts :: m OptionSet
+  makeSuccess :: String -> m r
+  makeFailure :: String -> m r
+
+getScriptResult ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m (Either ScriptError ([Text], ScriptResult))
+getScriptResult = do
+  script <- getScript
+  context <- getContext
+  args <- getTestData
+  pure $ case (script, args) of
+    (SomeSpender val, SpendingTest d r _) ->
+      let d' = Datum . toBuiltinData $ d
+          r' = Redeemer . toBuiltinData $ r
+       in testValidatorScript context val d' r'
+    (SomeMinter mp, MintingTest r) ->
+      let r' = Redeemer . toBuiltinData $ r
+       in testMintingPolicyScript context mp r'
+
+prepareConf :: OptionSet -> TransactionConfig
+prepareConf opts =
+  TransactionConfig
+    { testFee = testFee'
+    , testTimeRange = testTimeRange'
+    , testTxId = testTxId'
+    , testCurrencySymbol = testCurrencySymbol'
+    , testValidatorHash = testValidatorHash'
+    }
+  where
+    testFee' :: Value
+    Fee testFee' = lookupOption opts
+    testTimeRange' :: Interval POSIXTime
+    TimeRange testTimeRange' = lookupOption opts
+    testTxId' :: TxId
+    TestTxId testTxId' = lookupOption opts
+    testCurrencySymbol' :: CurrencySymbol
+    TestCurrencySymbol testCurrencySymbol' = lookupOption opts
+    testValidatorHash' :: ValidatorHash
+    TestValidatorHash testValidatorHash' = lookupOption opts
+
+getConf ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m TransactionConfig
+getConf = prepareConf <$> getOpts
+
+getScriptContext ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m ScriptContext
+getScriptContext = do
+  cb <- getCB
+  td <- getTestData
+  conf <- getConf
+  pure $ case td of
+    SpendingTest d _ v -> compileSpending conf cb d v
+    MintingTest _ -> compileMinting conf cb
+
+getContext ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m Context
+getContext = Context . toBuiltinData <$> getScriptContext
+
+getOpt ::
+  forall (v :: Type) (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  (IsOption v, MonadScriptTesting m p r) =>
+  m v
+getOpt = lookupOption <$> getOpts
+
+getShouldChat ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m PlutusTracing
+getShouldChat = getOpt
```

Comment:
This thing will be ignored in property tests, I think.

Reply:
Property tests don't currently deal with chattiness - the notion there would have to be rather different, since we perform shrinking.